File: scripts/ml_predictor.py
#!/usr/bin/env python3
"""
ML Predictor Helper
Loads trained model and makes single prediction
Called from Node.js mlPredictor.ts service

Input (stdin): JSON with {region, hour, latitude, longitude, month}
Output (stdout): JSON with {prediction: float}
"""

import sys
import json
from pathlib import Path

try:
    import joblib
    import pandas as pd
    import numpy as np
    from sklearn.preprocessing import LabelEncoder
except ImportError:
    # Missing dependencies
    print(json.dumps({"error": "Missing dependencies"}))
    sys.exit(1)

REPO_ROOT = Path(__file__).parent.parent
MODEL_PATH = REPO_ROOT / "models" / "xgb_model.bin"
METADATA_PATH = REPO_ROOT / "models" / "model_metadata.json"


def load_model_and_metadata():
    """Load trained model and metadata"""
    try:
        model = joblib.load(MODEL_PATH)
        with open(METADATA_PATH) as f:
            metadata = json.load(f)
        return model, metadata
    except Exception as e:
        print(json.dumps({"error": f"Failed to load model: {str(e)}"}))
        sys.exit(1)


def predict(model, metadata, input_data):
    """Make prediction with trained model"""
    try:
        region = input_data["region"]
        hour = int(input_data["hour"])
        latitude = float(input_data["latitude"])
        longitude = float(input_data["longitude"])
        month = int(input_data.get("month", 3))

        # Encode region
        region_classes = metadata["regions"]
        region_encoded = region_classes.index(region) if region in region_classes else 0

        # Prepare features (must match training order)
        features = [
            region_encoded,
            hour,
            month,
            latitude,
            longitude,
        ]

        # Convert to DataFrame
        X = pd.DataFrame([features], columns=metadata["feature_names"])

        # Predict
        prediction = float(model.predict(X)[0])

        # Clamp to reasonable range
        prediction = max(2, min(120, prediction))

        print(json.dumps({"prediction": prediction}))

    except Exception as e:
        print(json.dumps({"error": f"Prediction failed: {str(e)}"}))
        sys.exit(1)


if __name__ == "__main__":
    # Read input from stdin
    try:
        input_line = sys.stdin.read()
        input_data = json.loads(input_line)
    except json.JSONDecodeError:
        print(json.dumps({"error": "Invalid JSON input"}))
        sys.exit(1)

    # Load model and make prediction
    model, metadata = load_model_and_metadata()
    predict(model, metadata, input_data)

File: services/mlPredictor.ts
/**
 * ML Predictor Service
 * Loads trained XGBoost model and serves predictions
 *
 * Usage:
 *   await predictHaltDuration({ region: 'North', hour: 15, lat: 30.5, lng: 75.2 })
 *
 * Note: Requires Python environment with: pip install xgboost joblib
 * Model must be trained first: python scripts/train_model.py
 */

import * as fs from 'fs';
import * as path from 'path';
import { spawn } from 'child_process';
import { logger } from './logger';

interface PredictionInput {
  region: string;
  hour: number;
  latitude: number;
  longitude: number;
  month?: number;
}

interface PredictionResult {
  predicted_duration_min: number;
  confidence: number;
  method: 'ml' | 'fallback';
  model_info?: {
    r2_score: number;
    mae: number;
  };
}

const MODELS_DIR = path.join(process.cwd(), 'models');
const MODEL_PATH = path.join(MODELS_DIR, 'xgb_model.bin');
const METADATA_PATH = path.join(MODELS_DIR, 'model_metadata.json');

let modelMetadata: any = null;
let modelAvailable = false;

/**
 * Initialize ML service and load metadata
 */
export async function initializeMLP(): Promise<boolean> {
  try {
    if (fs.existsSync(METADATA_PATH)) {
      const metadataJson = fs.readFileSync(METADATA_PATH, 'utf-8');
      modelMetadata = JSON.parse(metadataJson);
      modelAvailable = true;
      logger.success('ML model loaded', {
        trained_at: modelMetadata.trained_at,
        r2_score: modelMetadata.r2_score,
      });
      return true;
    } else {
      logger.warn('ML model not found. Run: python scripts/train_model.py');
      modelAvailable = false;
      return false;
    }
  } catch (err) {
    logger.error('Failed to load ML model metadata:', err);
    modelAvailable = false;
    return false;
  }
}

/**
 * Call Python subprocess to make prediction
 * (Alternative to direct ML library loading in Node.js)
 */
function predictViaPython(input: PredictionInput): Promise<number | null> {
  return new Promise((resolve) => {
    const pythonScript = path.join(process.cwd(), 'scripts', 'ml_predictor.py');

    if (!fs.existsSync(pythonScript)) {
      logger.warn('[ML] Python predictor script not found');
      resolve(null);
      return;
    }

    const pythonProcess = spawn('python', [pythonScript], {
      timeout: 5000,
    });

    let output = '';
    let error = '';

    pythonProcess.stdout.on('data', (data) => {
      output += data.toString();
    });

    pythonProcess.stderr.on('data', (data) => {
      error += data.toString();
    });

    pythonProcess.on('close', (code) => {
      if (code === 0 && output) {
        try {
          const result = JSON.parse(output);
          resolve(result.prediction || null);
        } catch (err) {
          logger.debug('[ML] Failed to parse Python output:', err);
          resolve(null);
        }
      } else {
        if (error) logger.debug('[ML] Python error:', error);
        resolve(null);
      }
    });

    pythonProcess.on('error', (err) => {
      logger.debug('[ML] Python subprocess error:', err);
      resolve(null);
    });

    // Send input via stdin
    pythonProcess.stdin.write(JSON.stringify(input));
    pythonProcess.stdin.end();
  });
}

/**
 * Heuristic fallback prediction
 * Used when model not available
 */
function predictViaHeuristic(input: PredictionInput): number {
  let baseDuration = 5; // minutes

  // Regional factors
  const regionFactors: Record<string, number> = {
    North: 1.2,
    South: 0.9,
    East: 1.1,
    West: 0.8,
    Centre: 1.0,
  };

  const regionFactor = regionFactors[input.region] || 1.0;
  baseDuration *= regionFactor;

  // Hour of day factors (peak hours tend to have longer halts)
  const hourFactors: Record<number, number> = {
    6: 1.3,  // morning peak
    8: 1.5,
    9: 1.4,
    18: 1.4, // evening peak
    20: 1.2,
  };

  const hourFactor = hourFactors[input.hour] || 1.0;
  baseDuration *= hourFactor;

  // Add some randomness for realism
  const randomVariance = 0.8 + Math.random() * 0.4; // 0.8-1.2
  baseDuration *= randomVariance;

  return Math.round(Math.max(2, Math.min(60, baseDuration))); // Clamp 2-60 min
}

/**
 * Main prediction function
 */
export async function predictHaltDuration(
  input: PredictionInput
): Promise<PredictionResult> {
  const month = input.month || new Date().getMonth() + 1;

  // Validate input
  if (!input.region || input.hour < 0 || input.hour > 23) {
    logger.warn('[ML] Invalid prediction input:', input);
    return {
      predicted_duration_min: predictViaHeuristic(input),
      confidence: 0.3,
      method: 'fallback',
    };
  }

  // Try ML model first
  if (modelAvailable) {
    try {
      logger.debug('[ML] Using trained model for prediction:', {
        region: input.region,
        hour: input.hour,
      });

      // Attempt Python subprocess call
      const mlPrediction = await predictViaPython(input);

      if (mlPrediction !== null) {
        return {
          predicted_duration_min: Math.round(mlPrediction),
          confidence: Math.min(0.95, 0.7 + (modelMetadata.r2_score || 0) / 2),
          method: 'ml',
          model_info: {
            r2_score: modelMetadata.r2_score,
            mae: modelMetadata.mae,
          },
        };
      }
    } catch (err) {
      logger.debug('[ML] Model prediction failed, falling back to heuristic:', err);
    }
  }

  // Fallback to heuristic
  return {
    predicted_duration_min: predictViaHeuristic(input),
    confidence: 0.5,
    method: 'fallback',
  };
}

/**
 * Batch predictions for multiple trains
 */
export async function predictMultipleHalts(
  inputs: PredictionInput[]
): Promise<PredictionResult[]> {
  return Promise.all(inputs.map((input) => predictHaltDuration(input)));
}

/**
 * Get model status
 */
export function getModelStatus(): {
  available: boolean;
  metadata?: any;
} {
  if (!modelAvailable) {
    return {
      available: false,
    };
  }

  return {
    available: true,
    metadata: {
      trained_at: modelMetadata.trained_at,
      r2_score: modelMetadata.r2_score,
      training_samples: modelMetadata.training_samples,
      regions: modelMetadata.regions,
    },
  };
}

/**
 * Initialize on import
 */
initializeMLP().catch((err) => {
  logger.warn('[ML] Failed to initialize ML service:', err);
});
